docs/basic/complexity.md

```diff
@@ -1,101 +1,188 @@
+
+## Simple Examples of Calculating Time Complexity
+
+### for-expression
+
+```cpp
+// C++ Version
+int n, m;
+std::cin >> n >> m;
+for (int i = 0; i < n; ++i) {
+  for (int j = 0; j < n; ++j) {
+    for (int k = 0; k < m; ++k) {
+      std::cout << "hello world\n";
+    }
+  }
+}
+```
+
+```python
+# Python Version
+n = int(input())
+m = int(input())
+for i in range(0, n):
+    for j in range(0, n):
+        for k in range(0, m):
+            print("hello world")
+```
 
-## 主定理 (Master Theorem)
+The time complexity of the code above is $\Theta(n^2m)$ if we consider the inputted value of $n$ and $m$ as the data size. 
 
-我们可以使用 Master Theorem 来快速的求得关于递归算法的复杂度。
-假设我们有递推关系式
+### DFS
+
+While performing [depth-first search (article not translated)](../graph/dfs.md) on a graph with $n$ points and $m$ edges, as each point and edges will be accessed constant times, the complexity is $\Theta(n+m)$.
+
+## Identifying Constants
+
+When we are going to perform several operations, how to determine whether these "several operations" will impact time complexity? e.g.:
+
+```cpp
+// C++ Version
+const int N = 100000;
+for (int i = 0; i < N; ++i) {
+  std::cout << "hello world\n";
+}
+```
+
+```python
+# Python Version
+N = 100000
+for i in range(0, N):
+    print("hello world")
+```
+
+If the value of $N$ is not considered as data size of input, then time complexity of these blocks of code is $O(1)$.
+
+When calculating time complexity, it is important to figure out which variables should be treated as the size of data size. All other variables not relating to input data size will be treated as constants and treated as $1$ (or constant time) when calculating time complexity. 
+
+## Master Theorem
+
+By using [master theorem](https://en.wikipedia.org/wiki/Master_theorem_(analysis_of_algorithms)) we can quickly calculate the time complexity of a recursive algorithm.
+
+Assume we have a recurrence relation formula,
 
 $$
 T(n) = a T\left(\frac{n}{b}\right)＋f(n)\qquad \forall n > b
 $$
 
-那么
+Then, 
 
 $$
 T(n) = \begin{cases}\Theta(n^{\log_b a}) & f(n) = O(n^{\log_b a-\epsilon}) \\ \Theta(f(n)) & f(n) = \Omega(n^{\log_b a+\epsilon}) \\ \Theta(n^{\log_b a}\log^{k+1} n) & f(n)=\Theta(n^{\log_b a}\log^k n),k\ge 0 \end{cases}
 $$
 
-## 均摊复杂度
+## Amortized Complexity 
 
-算法往往是会对内存中的数据进行修改的，而同一个算法的多次执行，就会通过对数据的修改而互相影响。
+Algorithms tend to modify data in memory, and multiple executions of an algorithm will have impact on each other by modifying data.
 
-例如快速排序中的“按大小分类”操作，单次执行的最坏时间复杂度，看似是 $O(n)$ 的。
-但是由于快排的分治过程，先前的“分类”操作每次都减小了数组长度，所以实际的总复杂度 $O(n \log n)$，分摊在每一次“分类”操作上，是 $O(\log n)$。
+E.g., for the sorting-by-size operation in quicksort, the worst-case time complexity is seemed to be $O(n)$. But because of the divide-and-conquer process from quicksort, previous sorting operation reduces the length of the array every time. Therefore, the actual overall complexity is $O(n \log n)$, and the breakdowns over each sorting operation is $O(\log n)$. 
```

Comment:
```suggestion
E.g., for the sorting-by-size operation in quicksort, the worst-case time complexity is seemed to be $O(n)$. But because of the divide-and-conquer process from quicksort, previous sorting operation constantly reduces the length of the array. Therefore, the actual overall complexity is $O(n \log n)$, and the breakdowns over each sorting operation is $O(\log n)$. 
```
divide & conquer 可以 link 一下?

Reply:
原文没link